File: backend/app/validation.py
"""Gemeinsame Pruefungen fuer eingehende Daten der REST-Schnittstelle.

Jede Funktion prueft genau eine Sache. Bei Erfolg liefert sie den bereinigten
Wert zurueck, sonst wirft sie ValidationError. Die Ausnahme wird in
backend/app/__init__.py zentral in eine Antwort mit Status 400 uebersetzt.
"""

import re
from calendar import monthrange
from datetime import date

EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[A-Za-z]{2,}$")
TIME_PATTERN = re.compile(r"^([01][0-9]|2[0-3]):[0-5][0-9]$")

MAX_FUTURE_YEARS = 10


class ValidationError(Exception):
    """Eine Eingabe verletzt eine fachliche Regel."""

    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message = message


def _is_missing(value) -> bool:
    return value is None or (isinstance(value, str) and value.strip() == "")


def require_text(value, field_label: str, max_length: int) -> str:
    """Pflichttext: nicht leer, nicht laenger als die Datenbankspalte."""
    if _is_missing(value):
        raise ValidationError(f"{field_label} darf nicht leer sein")
    text = str(value).strip()
    if len(text) > max_length:
        raise ValidationError(f"{field_label} darf hoechstens {max_length} Zeichen lang sein")
    return text


def optional_text(value, field_label: str, max_length: int) -> str | None:
    if _is_missing(value):
        return None
    return require_text(value, field_label, max_length)


def require_email(value) -> str:
    email = require_text(value, "E-Mail-Adresse", 255).lower()
    if not EMAIL_PATTERN.match(email):
        raise ValidationError("E-Mail-Adresse muss die Form name@domain.de haben")
    return email


def require_password(value) -> str:
    if _is_missing(value):
        raise ValidationError("Passwort darf nicht leer sein")
    password = str(value)
    if len(password) < 6:
        raise ValidationError("Passwort muss mindestens 6 Zeichen haben")
    if len(password) > 128:
        raise ValidationError("Passwort darf hoechstens 128 Zeichen haben")
    return password


def require_int_in_range(
    value, field_label: str, minimum: int, maximum: int, default: int | None = None
) -> int:
    """Ganze Zahl im erlaubten Bereich.

    Fehlt der Wert und ist ein Standardwert angegeben, wird dieser benutzt.
    Die Zahl 0 gilt ausdruecklich als vorhandener Wert - sie darf nicht
    stillschweigend durch den Standardwert ersetzt werden.
    """
    if _is_missing(value):
        if default is not None:
            return default
        raise ValidationError(f"{field_label} ist ein Pflichtfeld")
    try:
        number = int(value)
    except (TypeError, ValueError):
        raise ValidationError(f"{field_label} muss eine ganze Zahl sein") from None
    if number < minimum or number > maximum:
        raise ValidationError(
            f"{field_label} muss zwischen {minimum} und {maximum} liegen"
        )
    return number


def require_future_date(value, field_label: str) -> date:
    """ISO-Datum (JJJJ-MM-TT), heute oder spaeter, hoechstens 10 Jahre voraus."""
    if _is_missing(value):
        raise ValidationError(f"{field_label} ist ein Pflichtfeld")
    try:
        parsed = date.fromisoformat(str(value))
    except ValueError:
        raise ValidationError(
            f"{field_label} muss im Format JJJJ-MM-TT angegeben werden"
        ) from None
    today = date.today()
    if parsed < today:
        raise ValidationError(f"{field_label} darf nicht in der Vergangenheit liegen")
    if parsed.year > today.year + MAX_FUTURE_YEARS:
        raise ValidationError(
            f"{field_label} darf hoechstens {MAX_FUTURE_YEARS} Jahre in der Zukunft liegen"
        )
    return parsed


def require_day_of_month(value, year: int, month: int) -> int | None:
    """Tag im Monat, passend zur Laenge des Monats (Schaltjahre eingeschlossen)."""
    if _is_missing(value):
        return None
    last_day = monthrange(year, month)[1]
    return require_int_in_range(value, "Tag des Monats", 1, last_day)


def optional_clock_time(value) -> str | None:
    """Uhrzeit im 24-Stunden-Format HH:MM, zum Beispiel 14:30."""
    if _is_missing(value):
        return None
    text = str(value).strip()
    if not TIME_PATTERN.match(text):
        raise ValidationError("Uhrzeit muss im Format HH:MM angegeben werden, zum Beispiel 14:30")
    return text

File: frontend/src/app/core/validation.ts
/**
 * Pruefregeln fuer Formulare. Jede Funktion liefert eine deutsche
 * Fehlermeldung oder null, wenn der Wert in Ordnung ist.
 *
 * Diese Regeln spiegeln bewusst die Pruefungen des Servers in
 * backend/app/validation.py. Der Server bleibt die verbindliche Instanz;
 * die Pruefung hier dient nur der schnellen Rueckmeldung im Browser.
 *
 * Hinweis: Die HTML-Attribute min/max/type="email" wirken in diesen
 * Formularen NICHT, weil Angular jedem ngModel-Formular das Attribut
 * novalidate hinzufuegt. Deshalb wird hier ausdruecklich geprueft.
 */
const EMAIL_PATTERN = /^[^@\s]+@[^@\s]+\.[A-Za-z]{2,}$/;
const TIME_PATTERN = /^([01][0-9]|2[0-3]):[0-5][0-9]$/;

export function validateEmail(value: string): string | null {
  const email = (value ?? '').trim();
  if (!email) return 'E-Mail-Adresse darf nicht leer sein';
  if (email.length > 255) return 'E-Mail-Adresse ist zu lang';
  if (!EMAIL_PATTERN.test(email)) return 'Bitte eine gültige Adresse angeben, z. B. name@domain.de';
  return null;
}

export function validatePassword(value: string): string | null {
  const password = value ?? '';
  if (password.length < 6) return 'Passwort muss mindestens 6 Zeichen haben';
  if (password.length > 128) return 'Passwort darf höchstens 128 Zeichen haben';
  return null;
}

export function validateRequiredText(value: string, label: string, maxLength = 255): string | null {
  const text = (value ?? '').trim();
  if (!text) return `${label} darf nicht leer sein`;
  if (text.length > maxLength) return `${label} darf höchstens ${maxLength} Zeichen lang sein`;
  return null;
}

export function validateEcts(value: number | null): string | null {
  if (value === null || value === undefined || Number.isNaN(value)) return 'ECTS-Punkte sind ein Pflichtfeld';
  if (!Number.isInteger(Number(value))) return 'ECTS-Punkte müssen eine ganze Zahl sein';
  if (Number(value) < 1 || Number(value) > 30) return 'ECTS-Punkte müssen zwischen 1 und 30 liegen';
  return null;
}

export function validateWorkloadHours(value: number | null): string | null {
  if (value === null || value === undefined || (value as unknown as string) === '') return null;
  const hours = Number(value);
  if (!Number.isInteger(hours)) return 'Lernaufwand muss eine ganze Zahl sein';
  if (hours < 1 || hours > 1000) return 'Lernaufwand muss zwischen 1 und 1000 Stunden liegen';
  return null;
}

export function validateTargetDate(value: string, current?: string): string | null {
  if (!value) return 'Zieldatum ist ein Pflichtfeld';
  const parsed = new Date(`${value}T00:00:00`);
  if (Number.isNaN(parsed.getTime())) return 'Zieldatum muss ein gültiges Datum sein';
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  if (parsed.getFullYear() > today.getFullYear() + 10) return 'Zieldatum liegt zu weit in der Zukunft';
  // Ein unveraendertes Datum bleibt zulaessig, auch wenn es verstrichen ist -
  // sonst liesse sich ein altes Lernziel nicht mehr bearbeiten.
  if (current && value === current) return null;
  if (parsed < today) return 'Zieldatum darf nicht in der Vergangenheit liegen';
  return null;
}

/** Tag im Monat; beruecksichtigt die Laenge des gewaehlten Monats. */
export function validateDayOfMonth(value: number | null, year: number, month: number): string | null {
  if (value === null || value === undefined || (value as unknown as string) === '') return null;
  const day = Number(value);
  if (!Number.isInteger(day)) return 'Tag muss eine ganze Zahl sein';
  const lastDay = new Date(year, month, 0).getDate();
  if (day < 1 || day > lastDay) return `Tag muss zwischen 1 und ${lastDay} liegen`;
  return null;
}

export function validateDuration(value: number | null): string | null {
  if (value === null || value === undefined || Number.isNaN(value)) return 'Dauer ist ein Pflichtfeld';
  const minutes = Number(value);
  if (!Number.isInteger(minutes)) return 'Dauer muss eine ganze Zahl sein';
  if (minutes < 5 || minutes > 480) return 'Dauer muss zwischen 5 und 480 Minuten liegen';
  return null;
}

export function validateClockTime(value: string): string | null {
  const text = (value ?? '').trim();
  if (!text) return null;
  if (!TIME_PATTERN.test(text)) return 'Uhrzeit muss im Format HH:MM angegeben werden, z. B. 14:30';
  return null;
}
